File: zivid_samples/src/sample_projection.cpp
#include <filesystem>
#include <rclcpp/rclcpp.hpp>
#include <std_srvs/srv/trigger.hpp>
#include <stdexcept>
#include <zivid_interfaces/srv/projection_resolution.hpp>
#include <zivid_interfaces/srv/projection_start.hpp>

static const auto read_only_parameter =
  rcl_interfaces::msg::ParameterDescriptor{}.set__read_only(true);

void fatal_error(const rclcpp::Logger & logger, const std::string & message)
{
  RCLCPP_ERROR_STREAM(logger, message);
  throw std::runtime_error(message);
}

auto create_projection_start_client(std::shared_ptr<rclcpp::Node> & node)
{
  auto client = node->create_client<zivid_interfaces::srv::ProjectionStart>("projection/start");
  while (!client->wait_for_service(std::chrono::seconds(3))) {
    if (!rclcpp::ok()) {
      fatal_error(node->get_logger(), "Client interrupted while waiting for service to appear.");
    }
    RCLCPP_INFO(node->get_logger(), "Waiting for the projection service to appear...");
  }

  RCLCPP_INFO(node->get_logger(), "projection service is available");
  return client;
}

auto create_resolution_client(std::shared_ptr<rclcpp::Node> & node)
{
  auto client =
    node->create_client<zivid_interfaces::srv::ProjectionResolution>("projection/resolution");
  while (!client->wait_for_service(std::chrono::seconds(3))) {
    if (!rclcpp::ok()) {
      fatal_error(node->get_logger(), "Client interrupted while waiting for service to appear.");
    }
    RCLCPP_INFO(node->get_logger(), "Waiting for the projection service to appear...");
  }

  RCLCPP_INFO(node->get_logger(), "projection service is available");
  return client;
}

auto create_stop_client(std::shared_ptr<rclcpp::Node> & node)
{
  auto client = node->create_client<std_srvs::srv::Trigger>("projection/stop");
  while (!client->wait_for_service(std::chrono::seconds(3))) {
    if (!rclcpp::ok()) {
      fatal_error(node->get_logger(), "Client interrupted while waiting for service to appear.");
    }
    RCLCPP_INFO(node->get_logger(), "Waiting for the projection service to appear...");
  }

  RCLCPP_INFO(node->get_logger(), "projection service is available");
  return client;
}

auto get_projection_resolution(std::shared_ptr<rclcpp::Node> & node)
{
  auto resolution_client = create_resolution_client(node);
  auto request = std::make_shared<zivid_interfaces::srv::ProjectionResolution::Request>();
  auto future = resolution_client->async_send_request(request);
  rclcpp::spin_until_future_complete(node, future);
  return future.get();
}

std::vector<uint8_t> generate_image(uint32_t width, uint32_t height)
{
  std::vector<uint8_t> pixel_data;
  pixel_data.reserve(width * height);

  for (size_t i = 0; i < height; i++) {
    for (size_t j = 0; j < width; j++) {
      const uint8_t blue = (256 * j) / width;
      const uint8_t red = (256 * (width - j)) / width;
      const uint8_t green = (256 * i) / height;
      pixel_data.emplace_back(blue);
      pixel_data.emplace_back(green);
      pixel_data.emplace_back(red);
      pixel_data.emplace_back(255);
    }
  }
  return pixel_data;
}

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = rclcpp::Node::make_shared("sample_projection");
  RCLCPP_INFO(node->get_logger(), "Started the sample_projection node");

  const auto image_path =
    node->declare_parameter<std::string>("image_path", "", read_only_parameter);

  auto resolution_client = create_resolution_client(node);
  auto projection_start_client = create_projection_start_client(node);
  auto stop_client = create_stop_client(node);

  RCLCPP_INFO(node->get_logger(), "Starting projection");
  auto request = std::make_shared<zivid_interfaces::srv::ProjectionStart::Request>();

  if (image_path.empty()) {
    const auto resolution = get_projection_resolution(node);
    request->data = generate_image(resolution->width, resolution->height);
  } else {
    request->image_path = image_path;
  }
  projection_start_client->async_send_request(request);

  RCLCPP_INFO(node->get_logger(), "Projecting for 5 seconds");
  rclcpp::spin_until_future_complete(
    node, std::promise<bool>().get_future(), std::chrono::seconds(5));

  RCLCPP_INFO(node->get_logger(), "Stopping projection");
  stop_client->async_send_request(std::make_shared<std_srvs::srv::Trigger::Request>());

  RCLCPP_INFO(node->get_logger(), "Spinning node.. Press Ctrl+C to abort.");
  rclcpp::spin(node);

  rclcpp::shutdown();
  return EXIT_SUCCESS;
}

File: zivid_samples/scripts/sample_projection.py
#!/usr/bin/env python

import sys

from rcl_interfaces.msg import ParameterDescriptor
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from std_srvs.srv import Trigger
from zivid_interfaces.srv import ProjectionResolution, ProjectionStart


def declare_and_get_parameter(node, name):
    value = (
        node.declare_parameter(name, '', ParameterDescriptor(read_only=True))
        .get_parameter_value()
        .string_value
    )
    return value


def generate_image(width, height):
    data = []
    for i in range(0, height):
        for j in range(0, width):
            blue = min(int((256 * j) / width), 255)
            red = min(int((256 * (width - j)) / width), 255)
            green = min(int((256 * i) / height), 255)
            data.extend([blue, green, red, 255])

    return data


class Sample(Node):

    def __init__(self):
        super().__init__('sample_projection')

        self.project_service = self.create_client(ProjectionStart, 'projection/start')
        while not self.project_service.wait_for_service(timeout_sec=3.0):
            self.get_logger().info('projection service not available, waiting again...')

        self.stop_service = self.create_client(Trigger, 'projection/stop')
        while not self.stop_service.wait_for_service(timeout_sec=3.0):
            self.get_logger().info('projection service not available, waiting again...')

        self.resolution_service = self.create_client(
            ProjectionResolution, 'projection/resolution'
        )
        while not self.project_service.wait_for_service(timeout_sec=3.0):
            self.get_logger().info('projection service not available, waiting again...')

    def get_resolution(self):
        self.get_logger().info('Calling resolution service')
        future = self.resolution_service.call_async(ProjectionResolution.Request())
        rclpy.spin_until_future_complete(self, future)
        return future.result()

    def project(self, image_path):
        if not image_path:
            r = self.get_resolution()
            request = ProjectionStart.Request(data=generate_image(r.width, r.height))
        else:
            request = ProjectionStart.Request(image_path=image_path)

        self.get_logger().info('Calling project service')
        return self.project_service.call_async(request)

    def stop(self):
        # self.get_logger().info('Calling stop service')
        # Cannot log; known issue: https://github.com/ros2/rclpy/issues/1287
        return self.stop_service.call_async(Trigger.Request())


def main(args=None):
    rclpy.init(args=args)

    try:
        sample = Sample()
        image_path = declare_and_get_parameter(sample, 'image_path')

        future = sample.project(image_path)
        rclpy.spin_until_future_complete(sample, future)

        sample.get_logger().info('Spinning node for 5 seconds')
        rclpy.spin_until_future_complete(sample, rclpy.task.Future(), timeout_sec=5)

        sample.get_logger().info('Stopping projection')
        future = sample.stop()
        rclpy.spin_until_future_complete(sample, future)

        sample.get_logger().info('Spinning node.. Press Ctrl+C to abort.')
        rclpy.spin(sample)

    except KeyboardInterrupt:
        pass
    except ExternalShutdownException:
        sys.exit(1)


if __name__ == '__main__':
    main()
